cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java

```diff
@@ -283,23 +342,27 @@ public void drop(Id.DatasetInstance instance) throws Exception {
    * @throws NamespaceNotFoundException if the requested namespace was not found
    * @throws IOException if there was a problem in checking if the namespace exists over HTTP
    */
-  public DatasetAdminOpResponse executeAdmin(Id.DatasetInstance instance, String method) throws Exception {
+  DatasetAdminOpResponse executeAdmin(Id.DatasetInstance instance, String method) throws Exception {
     // Throws NamespaceNotFoundException if the namespace does not exist
     ensureNamespaceExists(instance.getNamespace());
 
     Object result = null;
 
     // NOTE: one cannot directly call create and drop, instead this should be called thru
     //       POST/DELETE @ /data/datasets/{instance-id}. Because we must create/drop metadata for these at same time
+    Principal principal = SecurityRequestContext.toPrincipal();
     switch (method) {
       case "exists":
+        ensureAccess(instance.toEntityId());
         result = opExecutorClient.exists(instance);
         break;
       case "truncate":
+        authorizationEnforcementService.enforce(instance.toEntityId(), principal, Action.ADMIN);
         opExecutorClient.truncate(instance);
         publishAudit(instance, AuditType.TRUNCATE);
         break;
       case "upgrade":
+        authorizationEnforcementService.enforce(instance.toEntityId(), principal, Action.ADMIN);
```

Comment:
so this will be called by the Upgrade Tool. Will the user that runs the upgrade tool have admin privilege on all datasets?